DataDog/datadog-lambda-extension
# Generate CPU Enhanced Metrics

### What does this PR do?

This PR introduces eight new enhanced lambda metrics related to CPU usage. Each of the metrics are emitted once per invocation.

Three of the metrics represent CPU time spent running the lambda function:
- `aws.lambda.enhanced.cpu_system_time` - time spent by the CPU running in kernel mode
- `aws.lambda.enhanced.cpu_user_time` - time spent by the CPU running in user mode
- `aws.lambda.enhanced.cpu_total_time` - sum of `aws.lambda.enhanced.cpu_system_time` and `aws.lambda.enhanced.cpu_user_time`

The other five metrics represent CPU utilization by the lambda function:
- `aws.lambda.enhanced.cpu_total_utilization_pct` - total CPU utilization of the function, as a normalized percent (e.g. 35%)
- `aws.lambda.enhanced.cpu_total_utilization` - total CPU utilization of the function, as normalized cores (e.g. 1.2 cores)
- `aws.lambda.enhanced.num_cores` - total number of cores available in the environment (e.g. 4 cores)
- `aws.lambda.enhanced.cpu_min_utilization` - CPU utilization on the least utilized core, as a normalized percent (e.g. 10%)
- `aws.lambda.enhanced.cpu_max_utilization` - CPU utilization on the most utilized core, as a normalized percent (e.g. 80%)

#### Additional Notes
- The CPU time metrics are sent after the `PlatformReport` event (similar to how the network metrics are being sent) to match the metric values reported by Lambda Insights 
- The CPU utilization metrics are not based off Lambda Insights and were new metrics added to the go agent. These metrics are sent after the `PlatformRuntimeDone` event (similar to how these metrics are being sent in the go agent) to avoid accounting for additional idle time after this event. 
- The CPU utilization metrics are based on this metric format: [CPU Utilization Metric Representation](https://docs.google.com/document/d/1wpoaYlVM8ntHTsYTcYHFr3pCxq5l2_ynpp_Pg92LZqA/edit#heading=h.rfc7e5wnt2p)


### Testing

- [Build the lambda extension](https://datadoghq.atlassian.net/wiki/spaces/SLS/pages/2723611380/Lambda+Extension#Manual-Testing)
- Create a lambda function with the extension built in previous step
- Turn on lambda insights / add the [lambda insights extension](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Lambda-Insights-Getting-Started.html) to your function
- Invoke the function
- View the new enhanced metrics in metrics explorer or [this dashboard](https://ddserverless.datadoghq.com/dashboard/ker-x4u-kbr?fromUser=true&refresh_mode=sliding&view=spans&from_ts=1716303324812&to_ts=1716305124812&live=true)
- [Compare with lambda insight metrics](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Lambda-Insights-view-metrics.html) (for the CPU time metrics)
- Can verify the CPU utilization metrics by comparing the numbers of an invocation using the go agent vs. an invocation using the lambda extension to check that they are around the same range for that function
